emily/handler/src/database/accessors.rs

```diff
@@ -242,6 +209,174 @@ async fn wipe_chainstate_table(context: &EmilyContext) -> Result<(), Error> {
     .await
 }
 
+// Chainstate ------------------------------------------------------------------
+
+/// Add a chainstate entry.
+pub async fn add_chainstate_entry(
+    context: &EmilyContext,
+    entry: &ChainstateEntry,
+) -> Result<(), Error> {
+    // Get the existing chainstate entry for height. If there's a conflict
+    // then propagate it back to the caller.
+    let current_chainstate_entry_result = get_chainstate_entry_at_height(context, entry.key.height)
+        .await
+        .and_then(|existing_entry| {
+            if existing_entry != *entry {
+                Err(Error::InconsistentState(vec![
+                    entry.clone(),
+                    existing_entry,
+                ]))
+            } else {
+                Ok(())
+            }
+        });
+
+    // Exit here unless this chain height hasn't been detected before.
+    match current_chainstate_entry_result {
+        // Fall through if there is no existing entry..
+        Err(Error::NotFound) => (),
+        // ..otherwise exit here.
+        irrecoverable_or_okay => {
+            return irrecoverable_or_okay;
+        }
+    };
+
+    // TODO(390): Determine whether the order for these operations is correct
+    // given the eventual consistency guarantees of dynamodb.
+    //
+    // TODO(TBD): Handle api status being "Reorg" during this period.
+    let mut api_state = get_api_state(context).await?;
+    let blocks_higher_than_current_tip = entry.key.height - api_state.chaintip.key.height;
```

Comment:
How do we know that the `entry.key.height` is greater than the `api_state.chaintip.key.height`?

Reply:
Oh, given the else clause below, I think we want to cast these heights as `i64`s before doing the subtraction.

Reply:
Done.